src/lxml/includes/etree_defs.h

```diff
@@ -247,6 +247,22 @@ long _ftol2( double dblSource ) { return _ftol( dblSource ); }
 #define _isString(obj)   (PyUnicode_Check(obj) || PyBytes_Check(obj))
 #endif
 
+#if PY_MAJOR_VERSION < 3
+#define _isStringOrPathLike(obj)   (PyString_CheckExact(obj)  || \
+                                    PyUnicode_CheckExact(obj) || \
+                                    PyType_IsSubtype(Py_TYPE(obj), &PyBaseString_Type))
+#else
+#if PY_MINOR_VERSION < 6
+/* builtin subtype type checks are almost as fast as exact checks in Py2.7+
+ * and Unicode is more common in Py3 */
+#define _isStringOrPathLike(obj)   (PyUnicode_Check(obj) || PyBytes_Check(obj))
+#else
+#define _isStringOrPathLike(obj)   (PyUnicode_Check(obj) || \
+                                    PyBytes_Check(obj)   || \
+                                    PyObject_HasAttrString(obj, "__fspath__"))
```

Comment:
This is much better implemented in Cython with `hasattr()` in order to avoid creating a new string object on each attribute test. And, in fact, I'd rather have a Cython function that extracts the path directly if there is one, rather than testing for the attribute, then testing for the attribute again and then calling `os.fspath()` (which tests for the attribute again).

Additionally, if we avoid calling `os.fspath()`, then we can make also it work in Python < 3.6. There's a [`PyOS_FSPath()`](https://docs.python.org/3/c-api/sys.html#c.PyOS_FSPath) that we can call in Py3.6+, and you can implement a suitable fallback here (or in Cython) that just reads `obj.__fspath__`. One way to do it is to define `PyOS_FSPath` (or rather a safe alias) to `NULL` here for Py<3.6 and to avoid calling it based on `PY_VERSION_HEX` from Cython, using a separate simple Cython implementation in that case.

Reply:
I hope I understood your suggestions correctly. I now removed the `_isStringOrPathLike` function and added a cython function `_getFSPathOrObject`, which uses `PyOS_FSPath()` for python 3.6+ and gets `__fspath__()` for older python versions. This function I put before the `_isString` checks for the file arguments